common/src/main/java/com/proximyst/ban/service/MessageService.java

```diff
@@ -0,0 +1,230 @@
+        this.messageFactory.staticComponent(KEY_TARGET_UUID, target.getUuid().toString()));
+  }
+
+  public @NonNull IMessage commandsFeedbackUnban(final @NonNull BanUser target) {
+    return this.messageFactory.placeholderMessage(MessageKey.COMMANDS_FEEDBACK_UNBAN,
+        this.messageFactory.staticComponent(KEY_TARGET_NAME, target.getUsername()),
+        this.messageFactory.staticComponent(KEY_TARGET_UUID, target.getUuid().toString()));
+  }
+
+  public @NonNull IMessage commandsHistoryHeader(final @NonNull BanUser target, final @NonNegative int amount) {
+    return this.messageFactory.placeholderMessage(MessageKey.COMMANDS_HISTORY_HEADER,
+        this.messageFactory.staticComponent(KEY_TARGET_NAME, target.getUsername()),
+        this.messageFactory.staticComponent(KEY_TARGET_UUID, target.getUuid().toString()),
+        this.messageFactory.staticComponent(KEY_QUANTITY, Integer.toString(amount)));
+  }
+
+  public @NonNull IMessage commandsHistoryEntry(final @NonNull Punishment punishment) {
+    return this.messageFactory.placeholderMessage(MessageKey.COMMANDS_HISTORY_ENTRY,
+        this.messageFactory.componentComponent(this.createPlaceholders(punishment)));
+  }
+  // </editor-fold>
+
+  public @NonNull CompletableFuture<@NonNull Void> announceNewPunishment(final @NonNull Punishment punishment) {
+    return punishment.getPunishmentType()
+        .getBroadcastMessage(punishment.getReason().isPresent())
+        .map(key -> this.announcePunishment(punishment, key))
+        .orElseGet(() -> CompletableFuture.completedFuture(null));
+  }
+
+  public @NonNull CompletableFuture<@NonNull Void> announceLiftedPunishment(final @NonNull Punishment punishment) {
+    return punishment.getPunishmentType()
+        .getBroadcastLiftMessage()
+        .map(key -> this.announcePunishment(punishment, key))
+        .orElseGet(() -> CompletableFuture.completedFuture(null));
+  }
+
+  public @NonNull IMessage punishmentMessage(final @NonNull MessageKey messageKey,
+      final @NonNull Punishment punishment) {
+    return this.messageFactory.placeholderMessage(messageKey,
+        this.messageFactory.componentComponent(this.createPlaceholders(punishment)));
+  }
+
+  private @NonNull CompletableFuture<@NonNull Void> announcePunishment(final @NonNull Punishment punishment,
+      final @NonNull MessageKey messageKey) {
+    final String permission = punishment.getPunishmentType().getNotificationPermission().orElse(null);
+    return this.punishmentMessage(messageKey, punishment)
+        .component()
+        .thenAccept(component -> {
+          this.banServer.consoleAudience().sendMessage(Identity.nil(), component);
+          for (final IBanAudience audience : this.banServer.onlineAudiences()) {
+            if (permission != null && !audience.hasPermission(permission)) {
+              continue;
+            }
+
+            audience.sendMessage(Identity.nil(), component);
+          }
```

Comment:
Should be extracted into a new method. The for loop could also be converted into a cool Stream to be more uniform with the rest of the code.

Reply:
`@NonNull Iterable<? extends IBanAudience> onlineAudiences();` -- it cannot be streamed without some extra hacks I'm not willing to do.